File: src/milestones.ts
import { BADGE_NAMES, type PokemonStateObservation } from "./pokemon-state";
import type { MgbaObservation } from "./observation";

export const POKEMON_LEAGUE_MILESTONES = [
  "new-game-started",
  "player-control-reached",
  "first-battle-completed",
  "first-pokemon-obtained",
  "badge-1-boulder",   // Brock, Pewter City
  "badge-2-cascade",   // Misty, Cerulean City
  "badge-3-thunder",   // Lt. Surge, Vermilion City
  "badge-4-rainbow",   // Erika, Celadon City
  "badge-5-soul",      // Koga, Fuchsia City
  "badge-6-marsh",     // Sabrina, Saffron City
  "badge-7-volcano",   // Blaine, Cinnabar Island
  "badge-8-earth",     // Giovanni, Viridian City
  "all-badges-obtained",
  "victory-road-entered",
  "elite-four-started",
  "league-cleared",
] as const;

export type PokemonLeagueMilestoneId = (typeof POKEMON_LEAGUE_MILESTONES)[number];

const BADGE_MILESTONES: PokemonLeagueMilestoneId[] = [
  "badge-1-boulder",
  "badge-2-cascade",
  "badge-3-thunder",
  "badge-4-rainbow",
  "badge-5-soul",
  "badge-6-marsh",
  "badge-7-volcano",
  "badge-8-earth",
];

const milestoneRanks = new Map<PokemonLeagueMilestoneId, number>(
  POKEMON_LEAGUE_MILESTONES.map((m, i) => [m, i])
);

export interface MilestoneSnapshot {
  furthest: PokemonLeagueMilestoneId | null;
  badgesObtained: number;
  badgeNames: string[];
}

export class PokemonLeagueMilestoneTracker {
  #furthest: PokemonLeagueMilestoneId | null = null;
  #previousState: PokemonStateObservation | undefined;
  #previousBadges = 0;

  observe(observation: MgbaObservation): MilestoneSnapshot {
    const state = observation.state;
    if (state?.readStatus === "available") {
      this.#checkBadgeMilestones(state);
      this.#checkGameMilestones(state);
      this.#previousState = state;
      this.#previousBadges = state.badges;
    }
    return this.snapshot();
  }

  snapshot(): MilestoneSnapshot {
    const state = this.#previousState;
    const badges = state?.badges ?? this.#previousBadges;
    const badgeNames = BADGE_NAMES.filter((_, i) => (badges >> i) & 1);
    return {
      furthest: this.#furthest,
      badgesObtained: badgeNames.length,
      badgeNames,
    };
  }

  #checkBadgeMilestones(state: PokemonStateObservation): void {
    if (state.badgeCount === 0) return;

    // Check each badge bit
    for (let i = 0; i < 8; i++) {
      // biome-ignore lint/suspicious/noBitwiseOperators: badge bitmask check
      const hasNewBadge = (state.badges >> i) & 1;
      // biome-ignore lint/suspicious/noBitwiseOperators: badge bitmask check
      const hadBefore = (this.#previousBadges >> i) & 1;
      if (hasNewBadge && !hadBefore) {
        const milestone = BADGE_MILESTONES[i];
        if (milestone) this.#advance(milestone);
      }
    }

    if (state.badgeCount === 8) {
      this.#advance("all-badges-obtained");
    }
  }

  #checkGameMilestones(state: PokemonStateObservation): void {
    if (
      state.mapId === null ||
      state.position.x === null ||
      state.position.y === null
    ) {
      return;
    }

    if (this.#previousState?.battle && !state.battle) {
      this.#advance("first-battle-completed");
    }

    if (isPlayerControlState(state)) {
      this.#advance("player-control-reached");
    } else if (state.menuLike === true) {
      this.#advance("new-game-started");
    }
  }

  #advance(milestone: PokemonLeagueMilestoneId): void {
    if (isAfter(milestone, this.#furthest)) {
      this.#furthest = milestone;
    }
  }
}

function isPlayerControlState(state: PokemonStateObservation): boolean {
  return (
    !state.battle &&
    state.dialogueLike !== true &&
    state.menuLike !== true &&
    state.direction !== "unknown"
  );
}

function isAfter(
  candidate: PokemonLeagueMilestoneId,
  current: PokemonLeagueMilestoneId | null
): boolean {
  if (!current) return true;
  return (milestoneRanks.get(candidate) ?? -1) > (milestoneRanks.get(current) ?? -1);
}

File: backend/app/milestones.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional

from .observation import MgbaObservation
from .pokemon_state import PokemonState, BADGE_NAMES

MILESTONES = [
    "new-game-started",
    "player-control-reached",
    "first-battle-completed",
    "first-pokemon-obtained",
    "badge-1-boulder",
    "badge-2-cascade",
    "badge-3-thunder",
    "badge-4-rainbow",
    "badge-5-soul",
    "badge-6-marsh",
    "badge-7-volcano",
    "badge-8-earth",
    "all-badges-obtained",
    "victory-road-entered",
    "elite-four-started",
    "league-cleared",
]

_MILESTONE_RANK = {m: i for i, m in enumerate(MILESTONES)}

BADGE_MILESTONES = [
    "badge-1-boulder", "badge-2-cascade", "badge-3-thunder", "badge-4-rainbow",
    "badge-5-soul", "badge-6-marsh", "badge-7-volcano", "badge-8-earth",
]


@dataclass
class MilestoneSnapshot:
    furthest: str | None
    badges_obtained: int
    badge_names: list[str]


class MilestoneTracker:
    def __init__(self) -> None:
        self._furthest: str | None = None
        self._prev_state: PokemonState | None = None
        self._prev_badges = 0

    def observe(self, obs: MgbaObservation) -> MilestoneSnapshot:
        state = obs.state
        if state and state.read_status == "available":
            self._check_badge_milestones(state)
            self._check_game_milestones(state)
            self._prev_state = state
            self._prev_badges = state.badges
        return self.snapshot()

    def snapshot(self) -> MilestoneSnapshot:
        badges = self._prev_state.badges if self._prev_state else self._prev_badges
        badge_names = [n for i, n in enumerate(BADGE_NAMES) if (badges >> i) & 1]
        return MilestoneSnapshot(
            furthest=self._furthest,
            badges_obtained=len(badge_names),
            badge_names=badge_names,
        )

    def _check_badge_milestones(self, state: PokemonState) -> None:
        if state.badge_count == 0:
            return
        for i in range(8):
            if (state.badges >> i) & 1 and not ((self._prev_badges >> i) & 1):
                if i < len(BADGE_MILESTONES):
                    self._advance(BADGE_MILESTONES[i])
        if state.badge_count == 8:
            self._advance("all-badges-obtained")

    def _check_game_milestones(self, state: PokemonState) -> None:
        if state.map_id is None:
            return
        if self._prev_state and self._prev_state.battle and not state.battle:
            self._advance("first-battle-completed")
        if _is_player_control(state):
            self._advance("player-control-reached")
        elif state.menu_like:
            self._advance("new-game-started")

    def _advance(self, milestone: str) -> None:
        if _milestone_rank(milestone) > _milestone_rank(self._furthest):
            self._furthest = milestone


def _is_player_control(state: PokemonState) -> bool:
    return (
        not state.battle
        and not state.dialogue_like
        and not state.menu_like
        and state.direction != "unknown"
    )


def _milestone_rank(m: str | None) -> int:
    if m is None:
        return -1
    return _MILESTONE_RANK.get(m, -1)
